rfc/001-truncated-block-history.md

```diff
@@ -0,0 +1,87 @@
+
+The main configuration options involved are:
+
+* Unbonding time (SDK): the duration in which validators can be economically punished for misbehavior.
+
+* Block time (Tendermint): the minimum duration between consecutive blocks.
+
+* Block retention (Tendermint): the number of recent blocks (heights) to retain.
+
+* Snapshot interval (SDK): the interval (in heights) between taking state sync snapshots.
+
+* Snapshot retention (SDK): the number of recent state sync snapshots to retain.
+
+## Proposal
+
+* Unbonding time (SDK): already implemented as a genesis parameter `app_state.staking.params.unbonding_time`. Must now satisfy `unbonding_time < 2 * block.retention * block.time_iota_ms` otherwise errors (factor of 2 since `block.time_iota_ms` does not take into account block execution time).
+
+* Block time (Tendermint): already implemented as a genesis parameter `consensus_params.block.time_iota_ms`.
+
+* Block retention (Tendermint): implement as a new genesis parameter `consensus_params.block.retention` (default 0, i.e. retain all), and add a local configuration option `consensus.prune_blocks` (default off) that allows node operators to either enable or disable block pruning on a per-node basis. Must be at minimum 2, since the current and previous blocks are required for progress.
+
+* Snapshot interval (SDK): implement as a new genesis parameter `app_state.snapshots.interval` (default 0, i.e. disabled). Must satisfy `interval < 2 * block.retention` otherwise errors, to allow state synced nodes to catch up via block replay (factor of 2 to account for time spent taking a snapshot).
+
+* Snapshot retention (SDK): add a local configuration option `snapshot-enable` (default off) to enable/disable snapshots on a per-node basis, and `snapshot-retention` (default 2, 0 retains all) that specifies the number of snapshots to keep. Must be at least 2 to allow in-flight state syncs to complete as new snapshots are produced, but higher retention can be useful e.g. during mixed-version upgrades where old snapshots can be used by old-version nodes, or for state machine archives.
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+* Genesis parameters for block retention, block time, and unbonding time allows governance consensus to ensure availability and correctness of light client verification and short-term chain auditability, with enforced sanity checks.
+
+* Genesis parameters for snapshot interval ensures uniform snapshot schedules across nodes, providing high availability of snapshots at a given height for parallel downloads and faster state syncing, with enforced sanity checks.
+
+* Node operators can independently decide whether they want to provide complete block histories and snapshots.
+
+### Negative
+
+* Social coordination is required to run archival nodes, failure to do so may lead to permanent loss of historical blocks.
```

Comment:
Note I think it's possible for the chain to actually halt permanently too.

Reply:
How so?

Reply:
When all nodes restart and no nodes have blocks available to replay to the current state. Correct me if I'm wrong.

Reply:
Ah, this is something I wanted to look into. If it's the case that applications are not guaranteed to persist state on `Commit` then we need some signal from the app about what _is_ persisted, so that we don't prune blocks that haven't been persisted.

We don't currently seem to have this - on startup, Tendermint queries the last app height via `Query`, and then replays from there. However, at this point any transient data would be gone by definition (because it's after a restart). I think `Query` may return transient app heights during normal operation.

Personally I think `Commit` should be guaranteed to also persist data, but we could e.g. add `PersistentHeight` or `Persisted` in `CommitResponse` or something.

Reply:
This could actually be an alternative to the `block.retention` consensus parameter, as @marbar3778 suggested earlier. If the app returned `RetainBlocks` in `CommitResponse`, then it could take both persistence, unbonding time, and the snapshot interval into account and return an appropriate value.

Reply:
> Personally I think Commit should be guaranteed to also persist data

I also agree. The latest IAVL changes break this model of thinking for performance gains :-/

Using metadata via a Commit response could be promising. We should discuss on a call really the best approach forward.